ccl/changefeedccl: TestAlterChangefeedAddTargetFamily failed

ccl/changefeedccl.TestAlterChangefeedAddTargetFamily [failed](https://tanzanite.cluster.engflow.com/invocations/default/40ee6bae-8fa2-49e5-8f89-576e9d363f7e?testReportRun=3&testReportShard=3&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [d7ea85402dc35e36c6cc35520fa91f25fd5c999d](https://github.com/cockroachdb/cockroach/commits/d7ea85402dc35e36c6cc35520fa91f25fd5c999d):


```
=== RUN   TestAlterChangefeedAddTargetFamily
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestAlterChangefeedAddTargetFamily4088767951
    test_log_scope.go:76: use -show-logs to present logs inline
    alter_changefeed_test.go:275: -- test log scope end --
test logs left over in: outputs.zip/logTestAlterChangefeedAddTargetFamily4088767951
--- FAIL: TestAlterChangefeedAddTargetFamily (319.47s)
=== RUN   TestAlterChangefeedAddTargetFamily/kafka
    helpers_test.go:883: making server as secondary tenant
    helpers_test.go:962: making kafka feed factory
    alter_changefeed_test.go:266: 
        	Error Trace:	pkg/ccl/changefeedccl/helpers_test.go:198
        	            				pkg/ccl/changefeedccl/helpers_test.go:273
        	            				pkg/ccl/changefeedccl/alter_changefeed_test.go:266
        	            				pkg/ccl/changefeedccl/helpers_test.go:1134
        	            				pkg/ccl/changefeedccl/helpers_test.go:1169
        	Error:      	Received unexpected error:
        	            	operation "withTimeout-1027569918009311233" timed out after 5m0.005s (given timeout 5m0s): stopping job due to TestFeed timeout
        	            	(1) operation "withTimeout-1027569918009311233" timed out after 5m0.005s (given timeout 5m0s)
        	            	Wraps: (2) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.stopFeedWhenDone.func3
        	            	  | 	pkg/ccl/changefeedccl/testfeed_test.go:2957
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.stopFeedWhenDone.func1
        	            	  | 	pkg/ccl/changefeedccl/testfeed_test.go:2946
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (3) stopping job due to TestFeed timeout
        	            	Error types: (1) *timeutil.TimeoutError (2) *withstack.withStack (3) *errutil.leafError
        	Test:       	TestAlterChangefeedAddTargetFamily/kafka
    --- FAIL: TestAlterChangefeedAddTargetFamily/kafka (319.46s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=3</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAlterChangefeedAddTargetFamily.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-45343

ccl/changefeedccl.TestAlterChangefeedAddTargetFamily [failed](https://tanzanite.cluster.engflow.com/invocations/default/7ec3784f-893a-4ddb-9cc4-1b7cf3da080e?testReportRun=3&testReportShard=3&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [1aba4038242990fdb37a5317512fc5e82d6334fe](https://github.com/cockroachdb/cockroach/commits/1aba4038242990fdb37a5317512fc5e82d6334fe):


```
=== RUN   TestAlterChangefeedAddTargetFamily
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestAlterChangefeedAddTargetFamily1301895024
    test_log_scope.go:76: use -show-logs to present logs inline
    alter_changefeed_test.go:275: -- test log scope end --
test logs left over in: outputs.zip/logTestAlterChangefeedAddTargetFamily1301895024
--- FAIL: TestAlterChangefeedAddTargetFamily (319.89s)
=== RUN   TestAlterChangefeedAddTargetFamily/kafka
    helpers_test.go:883: making server as secondary tenant
    helpers_test.go:962: making kafka feed factory
    alter_changefeed_test.go:266: 
        	Error Trace:	pkg/ccl/changefeedccl/helpers_test.go:198
        	            				pkg/ccl/changefeedccl/helpers_test.go:273
        	            				pkg/ccl/changefeedccl/alter_changefeed_test.go:266
        	            				pkg/ccl/changefeedccl/helpers_test.go:1134
        	            				pkg/ccl/changefeedccl/helpers_test.go:1169
        	Error:      	Received unexpected error:
        	            	operation "withTimeout-1036356072082669569" timed out after 5m0s (given timeout 5m0s): stopping job due to TestFeed timeout
        	            	(1) operation "withTimeout-1036356072082669569" timed out after 5m0s (given timeout 5m0s)
        	            	Wraps: (2) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.stopFeedWhenDone.func3
        	            	  | 	pkg/ccl/changefeedccl/testfeed_test.go:2987
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.stopFeedWhenDone.func1
        	            	  | 	pkg/ccl/changefeedccl/testfeed_test.go:2976
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (3) stopping job due to TestFeed timeout
        	            	Error types: (1) *timeutil.TimeoutError (2) *withstack.withStack (3) *errutil.leafError
        	Test:       	TestAlterChangefeedAddTargetFamily/kafka
    --- FAIL: TestAlterChangefeedAddTargetFamily/kafka (319.89s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=3</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAlterChangefeedAddTargetFamily.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/ccl/changefeedccl/alter_changefeed_test.go
+++ b/pkg/ccl/changefeedccl/alter_changefeed_test.go
@@ -246,13 +246,30 @@ func TestAlterChangefeedAddTargetFamily(t *testing.T) {
 		testFeed := feed(t, f, `CREATE CHANGEFEED FOR foo FAMILY onlya`)
 		defer closeFeed(t, testFeed)
 
-		sqlDB.Exec(t, `INSERT INTO foo VALUES(1, 'hello')`)
-		assertPayloads(t, testFeed, []string{
-			`foo.onlya: [1]->{"after": {"a": 1}}`,
-		})
-
 		feed, ok := testFeed.(cdctest.EnterpriseTestFeed)
 		require.True(t, ok)
+
+		tsr := sqlDB.QueryRow(t, `INSERT INTO foo VALUES(42, 'hello') RETURNING cluster_logical_timestamp()`)
+		var insertTsDecStr string
+		tsr.Scan(&insertTsDecStr)
+		insertTs := parseTimeToHLC(t, insertTsDecStr)
+		assertPayloads(t, testFeed, []string{
+			`foo.onlya: [42]->{"after": {"a": 42}}`,
+		})
+
+		// Wait for the high water mark (aka resolved ts) to advance past the row we inserted's
+		// mvcc ts. Otherwise, we'd see [42] again due to a catch up scan, and it
+		// would muddy the waters.
+		testutils.SucceedsSoon(t, func() error {
+			registry := s.Server.JobRegistry().(*jobs.Registry)
+			job, err := registry.LoadJob(context.Background(), feed.JobID())
+			require.NoError(t, err)
+			prog := job.Progress()
+			if p := prog.GetHighWater(); p != nil && !p.IsEmpty() && insertTs.Less(*p) {
+				return nil
+			}
+			return errors.New("waiting for highwater")
+		})
 
 		sqlDB.Exec(t, `PAUSE JOB $1`, feed.JobID())
 		waitForJobStatus(sqlDB, t, feed.JobID(), `paused`)
@@ -261,12 +278,12 @@ func TestAlterChangefeedAddTargetFamily(t *testing.T) {
 
 		sqlDB.Exec(t, fmt.Sprintf(`RESUME JOB %d`, feed.JobID()))
 		waitForJobStatus(sqlDB, t, feed.JobID(), `running`)
-
-		sqlDB.Exec(t, `INSERT INTO foo VALUES(2, 'goodbye')`)
+		sqlDB.Exec(t, `INSERT INTO foo VALUES(37, 'goodbye')`)
 		assertPayloads(t, testFeed, []string{
-			`foo.onlyb: [1]->{"after": {"b": "hello"}}`,
-			`foo.onlya: [2]->{"after": {"a": 2}}`,
-			`foo.onlyb: [2]->{"after": {"b": "goodbye"}}`,
+			// Note that we don't see foo.onlyb.[42] here, because we're not
+			// doing a catchup scan and we've already processed that tuple.
+			`foo.onlya: [37]->{"after": {"a": 37}}`,
+			`foo.onlyb: [37]->{"after": {"b": "goodbye"}}`,
 		})
 	}
 
@@ -287,13 +304,31 @@ func TestAlterChangefeedSwitchFamily(t *testing.T) {
 		testFeed := feed(t, f, `CREATE CHANGEFEED FOR foo FAMILY onlya`)
 		defer closeFeed(t, testFeed)
 
-		sqlDB.Exec(t, `INSERT INTO foo VALUES(1, 'hello')`)
+		tsr := sqlDB.QueryRow(t, `INSERT INTO foo VALUES(1, 'hello') RETURNING cluster_logical_timestamp()`)
+		var insertTsDecStr string
+		tsr.Scan(&insertTsDecStr)
+		insertTs := parseTimeToHLC(t, insertTsDecStr)
+
 		assertPayloads(t, testFeed, []string{
 			`foo.onlya: [1]->{"after": {"a": 1}}`,
 		})
 
 		feed, ok := testFeed.(cdctest.EnterpriseTestFeed)
 		require.True(t, ok)
+
+		// Wait for the high water mark (aka resolved ts) to advance past the row we inserted's
+		// mvcc ts. Otherwise, we'd see [1] again due to a catch up scan, and it
+		// would muddy the waters.
+		testutils.SucceedsSoon(t, func() error {
+			registry := s.Server.JobRegistry().(*jobs.Registry)
+			job, err := registry.LoadJob(context.Background(), feed.JobID())
+			require.NoError(t, err)
+			prog := job.Progress()
+			if p := prog.GetHighWater(); p != nil && !p.IsEmpty() && insertTs.Less(*p) {
+				return nil
+			}
+			return errors.New("waiting for highwater")
+		})
 
 		sqlDB.Exec(t, `PAUSE JOB $1`, feed.JobID())
 		waitForJobStatus(sqlDB, t, feed.JobID(), `paused`)
@@ -305,7 +340,8 @@ func TestAlterChangefeedSwitchFamily(t *testing.T) {
 
 		sqlDB.Exec(t, `INSERT INTO foo VALUES(2, 'goodbye')`)
 		assertPayloads(t, testFeed, []string{
-			`foo.onlyb: [1]->{"after": {"b": "hello"}}`,
+			// Note that we don't see foo.onlyb.[1] here, because we're not
+			// doing a catchup scan and we've already processed that tuple.
 			`foo.onlyb: [2]->{"after": {"b": "goodbye"}}`,
 		})
 	}

